src/engines/kotor/gui/options/resolution.cpp

```diff
@@ -22,23 +22,64 @@
  *  The graphics resolution menu.
  */
 
+#include <SDL_video.h>
+#include "src/common/strutil.h"
+#include "src/common/ustring.h"
+
+#include "src/graphics/windowman.h"
+
 #include "src/engines/kotor/gui/options/resolution.h"
 
 #include "src/engines/aurora/kotorjadegui/panel.h"
+#include "src/engines/aurora/kotorjadegui/listbox.h"
 
 namespace Engines {
 
 namespace KotOR {
 
-OptionsResolutionMenu::OptionsResolutionMenu(Console *console) : GUI(console) {
+OptionsResolutionMenu::OptionsResolutionMenu(Console *console) : GUI(console), _newWidth(0), _newHeight(0) {
 	load("optresolution");
 
 	WidgetPanel *guiPanel = getPanel("TGuiPanel");
 	guiPanel->setPosition(-guiPanel->getWidth()/2, -guiPanel->getHeight()/2, 0);
+
+	int currentIndex = -1;
+	std::vector<Graphics::DisplayMode> modes = WindowMan.getDisplayModes();
+
+	// Filter every resolution smaller than 800x600
+	for (size_t i = 0; i < modes.size(); ++i) {
+		bool duplicate = false;
+		for (size_t j = 0; j < _modes.size(); ++j) {
+			if (_modes[j].w == modes[i].w && _modes[j].h == modes[i].h)
+				duplicate = true;
+		}
+
+		if (modes[i].w >= 800 && modes[i].h >= 600 && !duplicate) {
+			_modes.push_back(modes[i]);
+		}
+	}
+
+	WidgetListBox *listBox = getListBox("LB_RESOLUTIONS");
+	listBox->createItemWidgets(_modes.size());
+	for (size_t i = 0; i < _modes.size(); ++i) {
+		Graphics::DisplayMode mode = _modes[i];
+		listBox->addItem(Common::composeString(mode.w) + " x " + Common::composeString(mode.h));
+	}
+
+	listBox->setAdjustHeight(true);
+	listBox->setItemSelectionEnabled(true);
+	listBox->refreshItemWidgets();
 }
 
 void OptionsResolutionMenu::callbackActive(Widget &widget) {
-	// TODO: Add changing of resolutions
+	if (widget.getTag() == "BTN_OK") {
```

Comment:
You also need to update the "width" and "height" config variables here, so that a resolution change is saved in the config file and sticks with the next restart of xoreos.